packages/client/src/connection.js

```diff
@@ -41,12 +40,44 @@ class Connection {
  * @template {Record<string, any>} T
  * @template {API.Tuple<API.ServiceInvocation<C, T>>} I
  * @param {API.Connection<T>} connection
- * @param {I} invocations
- * @returns {Promise<API.InferServiceInvocations<I, T>>}
+ * @param {I} workflow
+ * @returns {Promise<API.InferWorkflowReceipts<I, T>>}
  */
-export const execute = async (invocations, connection) => {
-  const request = await connection.encoder.encode(invocations, connection)
+export const execute = async (workflow, connection) => {
+  const request = await connection.codec.encode(workflow, connection)
   const response = await connection.channel.request(request)
-  const result = await connection.decoder.decode(response)
-  return result
+  // We may fail to decode the response if content type is not supported
+  // or if data was corrupted. We do not want to throw in such case however,
+  // because client will get an Error object as opposed to a receipt, to retain
+  // consistent client API with two kinds of errors we encode caught error as
+  // a receipts per workflow invocation.
+  try {
+    return await connection.codec.decode(response)
+  } catch (error) {
+    const { message, ...cause } = /** @type {Error} */ (error)
+    const receipts = []
+    for await (const invocation of workflow) {
+      const { cid } = await invocation.delegate()
+      const receipt = await Receipt.issue({
```

Comment:
I agree this is not ideal, yet I feel very strongly that two channels for error handling is going to be a super error prone. 

```ts
try {
   const receipt = await agent.invoke(capbility)
    if (receipt.out.error) {
         // handle error
    }
} catch (error) {
   // handle error
}
```

Perhaps we could come up with a better solution so that they are separate steps, but lets do it in a followup